File: c/chapitre_4/exo44Pere.c
/*
ême question qu’à l’exercice ?? mais en passant les descripteurs de tube
comme variables d’environnement.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>

int main(int argc , char **argv)
{
    pid_t pid ;
    int tube[2] ;
    char varEnvName[100] ;
    char _read[10] , _write[10] ;

    if (argc != 2)
    {
        fprintf(stderr , "Usage : %s <Chemin_Program>\n", argv[0] ) ;
        exit(1);
    }
    
    // Creation du pipe
    if ( pipe(tube) == -1 )
    {
        perror("Erreur pipe\n");
        exit(1);
    }

    // Creation du varible d'environement
    sprintf( varEnvName , "DESCRIPTEUR=%d:%d" , tube[0] , tube[1] ) ;
    putenv(varEnvName);

    // Lancement du fork
    switch (pid=fork() )
    {
    case -1 :
        perror("Erreur Fork\n");
        exit(0);
        break;
    case 0 :
        char *args[] = {argv[1] , "DESCRIPTEUR" , NULL } ;
        execv(argv[1] , args ) ;

        perror("Erreur execv\n");
        exit(1) ;

        break;
    default:

        // Envoie du message du Père
        close(tube[0]);
        char *message = "Salut je suis ton Père !!\n" ;

        write(tube[1] , message , strlen(message)+1 ) ;

        close(tube[1]);

        wait(NULL) ;
        break;
    }
    return (0) ;

}

File: python/chapitre_4/exo44Pere.py
"""
ême question qu'à l'exercice ?? mais en passant les descripteurs de tube
comme variables d'environnement.
"""

import os
import sys

def main():
    if len(sys.argv) != 2:
        print(f"Usage : {sys.argv[0]} <Chemin_Program>", file=sys.stderr)
        sys.exit(1)

    # Creation du pipe
    tube_r, tube_w = os.pipe()

    # Creation du variable d'environement
    os.environ["DESCRIPTEUR"] = f"{tube_r}:{tube_w}"

    # Lancement du fork
    try:
        pid = os.fork()
    except OSError:
        print("Erreur Fork")
        sys.exit(0)

    if pid == 0:
        args = [sys.argv[1], "DESCRIPTEUR"]
        os.execv(sys.argv[1], args)

        print("Erreur execv")
        sys.exit(1)

    else:
        # Envoie du message du Père
        os.close(tube_r)
        message = "Salut je suis ton Père !!\n"
        os.write(tube_w, message.encode() + b'\x00')

        os.close(tube_w)
        os.wait()

    return 0

main()